### ISSUE TITLE
Refactoring of client?

### ISSUE THREAD

**Issue by jekyc** [OpenCTI-Platform/client-python#14]
The client is currently implemented by two primary classes, both of which are rather large (approx. 3600 and 1200 lines of code) and depend on each other. Would it make sense to split the client up? This would make it easier for developers to use and extend the client.

One implementation could be to have individual objects for each of the types used by OpenCTI.  These would be responsible for converting corresponding stix2 objects (from the python library stix2) to something the graphql api understands, and adding and validating OpenCTI custom properties.
Futhermore, these objects could also include the graphql queries for the object type (such as 'create', 'create_if_not_exists', etc).

Using this approach, the Identity class would look something like this:

```python3
class OpenCTIBaseModel:
    """Base model for OpenCTI objects

    Contains connection info for the graphql db and generic queries + misc helpers.
    """


class Identity(OpenCTIBaseModel):
    STIX_TYPE = stix2.Identity

    def __init__(self, *args, **kwargs):
        self.type = kwargs.get('type', None)
        self.name = kwargs.get('name', None)
        self.description = kwargs.get('description', None)
        self.id = kwargs.get('id', None)
        self.stix_id = kwargs.get('stix_id', None)
        self.created = kwargs.get('created', None)
        self.modified = kwargs.get('modified', None)

    @classmethod
    def get(cls, id):
        query = """
            query Identity($id: String!) {
                identity(id: $id) {
                   id
                    entity_type
                    stix_id
                    stix_label
                    name
                    alias
                    description
                    created
                    modified
                    createdByRef {
                        node {
                            id
                            entity_type
                            stix_id
                            stix_label
                            name
                            alias
                            description
                            created
                            modified
                        }
                    }
                    markingDefinitions {
                        edges {
                            node {
                                id
                                entity_type
                                stix_id
                                definition_type
                                definition
                                level
                                color
                                created
                                modified
                            }
                        }
                    }
                }
            }
        """
	result = cls.query(query, {'id': id})
	return cls.from_query(result)

    @classmethod
    def get_many(cls, limit=10000):
        query = """
            query Identities($first: Int) {
                identities(first: $first) {
                    edges {
                        node {
                            id
                            entity_type
                            stix_id
                            stix_label
                            name
                            alias
                            description
                            created
                            modified
                            createdByRef {
                                node {
                                    id
                                    entity_type
                                    stix_id
                                    stix_label
                                    name
                                    alias
                                    description
                                    created
                                    modified
                                }
                            }
                            markingDefinitions {
                                edges {
                                    node {
                                        id
                                        entity_type
                                        stix_id
                                        definition_type
                                        definition
                                        level
                                        color
                                        created
                                        modified
                                    }
                                }
                            }
                        }
                    }
                }
            }
        """
        result = self.query(query, {'first': limit})
	return [cls(**identity) for identity in result['data']['identities']]

    @classmethod
    def exists(cls, id=None, name=None):
        # implement actual exists query here or use the OpenCTIBaseModel generic queries such as (check_existing_stix_domain_entity)
        pass

    @classmethod
    def from_stix2(cls, obj):
        if isinstance(obj, cls.STIX_TYPE):
            return cls(**obj.serialize())

        return None

    @classmethod
    def from_query(cls, query_result):
        # this most likely needs some transformation of query_result...
        return cls(query_result)

    def save(self, only_if_not_exists=True):
        # query for saving the object
        pass

    def to_stix2(self):
        # this needs some more data mangling
        return cls.STIX_TYPE(**self, allow_custom=True)

```

With this approach it is easy to add custom properties based on the type of object - as an example the custom properties 'x_opencti_observable_type' and 'x_opencti_observable_value' for indicator/observable could be automatically set by parsing the STIX indicator pattern (assuming it consists of only one value). Another could be to automatically generate external references for Vulnerabilities (as requested in https://github.com/OpenCTI-Platform/opencti/issues/147). These custom changes/additions would be be easy to spot, maintain and document.

The client would then be some convenience methods (such as 'import_bundle'), that can convert the input to these OpenCTI objects and save them using the objects' save methods.

What do you think about this approach?

**Comment by SamuelHassine** [comment]
@jekyc: The OpenCTI Python client was created for helping people to better understand the underlying concepts of OpenCTI and its GraphQL API. We started using it in workers and so we needed to implement all the methods. All the library need to be refactored so thank you very much for your ideas.

Your approach is very close of what we intend to do in the future, the use of `kwarg**` and the general approach of one class by entity type, with generic OpenCTI for helper. 

We are glad you propose this kind of refactor. Perfect catch!
